packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts

```diff
@@ -42,44 +45,64 @@ export interface UserPoolClientProps {
 
   /**
    * List of enabled authentication flows
-   * @default no enabled flows
+   * @default - no enabled flows
    */
   readonly enabledAuthFlows?: AuthFlow[]
 }
 
 /**
- * Define a UserPool App Client
+ * Represents a Cognito user pool client.
  */
-export class UserPoolClient extends Resource {
+export interface IUserPoolClient extends IResource {
   /**
+   * Name of the application client
    * @attribute
    */
-  public readonly userPoolClientId: string;
+  readonly userPoolClientId: string;
+}
 
+/**
+ * Define a UserPool App Client
+ */
+export class UserPoolClient extends Resource implements IUserPoolClient {
   /**
-   * @attribute
+   * Import a user pool client given its id.
    */
-  public readonly userPoolClientName: string;
+  public static fromUserPoolClientId(scope: Construct, id: string, userPoolClientId: string): IUserPoolClient {
+    class Import extends Resource implements IUserPoolClient {
+      public readonly userPoolClientId = userPoolClientId;
+    }
+
+    return new Import(scope, id);
+  }
+
+  public readonly userPoolClientId: string;
 
   /**
-   * @attribute
+   * The client name that was specified via the `userPoolClientName` property during initialization,
+   * `undefined` otherwise.
+   */
+  public readonly userPoolClientName?: string;
```

Comment:
I would expect this to be an attribute and never `undefined`, no?

Reply:
No, does not work. CloudFormation returns the message "Name attribute not supported at this time, please use the CLI or Console to retrieve this value"

See my comment here - https://github.com/aws/aws-cdk/pull/7091/files#diff-112b6222454a4f23b78d56cd0fa66b95R88-R92

Reply:
Then I would implement this as a getter that throws an error if the name is not provided but at least we get a clean api that we can evolve later to return a token.

Reply:
Good point